mixer/adapter/kubernetesenv/kubernetesenv.go

```diff
@@ -123,32 +129,122 @@ func (b *builder) Build(ctx context.Context, env adapter.Env) (adapter.Handler,
-		env.Logger().Infof("Waiting for kubernetes cache sync...")
-		if success := cache.WaitForCacheSync(stopChan, controller.HasSynced); !success {
-			stopChan <- struct{}{}
-			return nil, errors.New("cache sync failure")
+
+		controller, err = getNewCacheController(b, clientset, env)
+		if err != nil {
+			return nil, fmt.Errorf("could not create new cache controller: %v", err)
 		}
-		env.Logger().Infof("Cache sync successful.")
+
+		controllers = append(controllers, controller)
 		b.controllers[path] = controller
+
+		remoteControllers, err = createRemoteCacheControllers(b, clientset, env)
+		if err == nil {
+			if len(remoteControllers) != 0 {
+				controllers = append(controllers, remoteControllers...)
+			}
+		} else {
+			return nil, fmt.Errorf("failure on creating remote controllers: %v", err)
+		}
+	} else {
+		for key := range b.controllers {
+			controllers = append(controllers, b.controllers[key])
+		}
 	}
 
+	env.Logger().Infof("installed %d controllers", len(controllers))
+
 	return &handler{
 		env:      env,
-		k8sCache: controller,
+		k8sCache: controllers,
 		params:   paramsProto,
 	}, nil
 }
 
+func createRemoteCacheControllers(b *builder, clientset k8s.Interface,
+	env adapter.Env) ([]cacheController, error) {
+	var controller cacheController
+	var k8sInterface k8s.Interface
+	var remoteControllers []cacheController
+	var opts meta_v1.ListOptions
+
+	secretNamespace := "istio-system" // TODO: make configurable
+	mcLabel := "istio/multiCluster"
+	opts.LabelSelector = mcLabel + "=true"
+
+	kubeSecrets, err := clientset.CoreV1().Secrets(secretNamespace).List(opts)
+	if err != nil {
+		return nil, fmt.Errorf("could not access secrets for namespace: %s error: %v",
+			secretNamespace, err)
+	}
+
+	for _, secret := range kubeSecrets.Items {
+		for dataKey, kubeconfig := range secret.Data {
+			k8sInterface, err = getK8sInterface(kubeconfig)
+			if err != nil {
+				return nil, fmt.Errorf("error on K8s interface access: %v", err)
+			}
+
+			controller, err = getNewCacheController(b, k8sInterface, env)
+			if err == nil {
+				remoteControllers = append(remoteControllers, controller)
+				b.controllers[dataKey] = controller
+			} else {
+				return nil, fmt.Errorf("error on creating remote controller: %s", dataKey)
+			}
+		}
+	}
+
+	return remoteControllers, nil
+}
+
+func createK8sInterface(kubeconfig []byte) (k8s.Interface, error) {
+	config, err := clientcmd.Load(kubeconfig)
+	if err != nil {
+		return nil, fmt.Errorf("could not load kubeconfig for secret error: %v", err)
+	}
+
+	clientConfig := clientcmd.NewDefaultClientConfig(*config, &clientcmd.ConfigOverrides{})
+	rest, err := clientConfig.ClientConfig()
+	if err != nil {
+		return nil, fmt.Errorf("error on ClientConfig access: %v", err)
+	}
+
+	k8sInterface, err := k8s.NewForConfig(rest)
+	if err != nil {
+		return nil, fmt.Errorf("error on NewforConfig access: %v", err)
+	}
+
+	return k8sInterface, nil
+}
+
+func getNewCacheController(b *builder, clientset k8s.Interface,
```

Comment:
style nit: why is this line wrapped ?

Reply:
Mostly because I come from other projects that restrict column length to 80. I know that Go has no restrictions and I haven't found any Istio code style that dictates line length (is there one?). From golang.org, I found "Go has no line length limit. Don't worry about overflowing a punched card. If a line feels too long, wrap it and indent with an extra tab. " So from this statement, (1) the person that wrote it is old enough to remember punch cards (as am I) and (2) break a line were ever you see fit. 

This isn't my file so I have no problems not breaking these lines to keep below 90. In this case the 'func' line would be 107 columns.